File: meme-generator-backend/app/src/main/java/com/example/data/api/GeminiClient.kt
package com.example.data.api

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import android.util.Log
import com.example.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Client unique pour toutes les interactions avec l'API Google Gemini.
 *
 * 100% GRATUIT — Aucune dépendance OpenAI.
 *
 * Modèles utilisés :
 *   gemini-2.0-flash             → texte, vision, audio  (1 500 req/jour gratuit)
 *   gemini-2.5-flash-preview-05-20 → génération d'image  (quota limité gratuit)
 *
 * Clé API : https://aistudio.google.com/app/apikey
 */
object GeminiClient {

    private const val TAG = "GeminiClient"
    private const val BASE_URL = "https://generativelanguage.googleapis.com"

    private const val MODEL_TEXT      = "gemini-2.0-flash"
    private const val MODEL_IMAGE_GEN = "gemini-2.5-flash-preview-05-20"

    private val client = OkHttpClient.Builder()
        .connectTimeout(60, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .build()

    private fun getApiKey(): String = BuildConfig.GEMINI_API_KEY

    private fun isKeyMissing(key: String) = key.isEmpty() || key == "MY_GEMINI_API_KEY"

    // ─────────────────────────────────────────────────────────
    // HELPERS PRIVÉS
    // ─────────────────────────────────────────────────────────

    /** Corps JSON Gemini pour une requête texte simple. */
    private fun buildTextBody(prompt: String): JSONObject = JSONObject().apply {
        put("contents", JSONArray().apply {
            put(JSONObject().apply {
                put("parts", JSONArray().apply {
                    put(JSONObject().apply { put("text", prompt) })
                })
            })
        })
        put("generationConfig", JSONObject().apply {
            put("responseMimeType", "application/json")
        })
    }

    /**
     * Exécute un POST vers Gemini et retourne le texte brut de la réponse.
     * Retourne null en cas d'erreur HTTP ou de réponse vide.
     */
    private fun callGemini(model: String, body: JSONObject, apiKey: String): String? {
        return try {
            val requestBody = body.toString().toRequestBody("application/json".toMediaType())
            val request = Request.Builder()
                .url("$BASE_URL/v1beta/models/$model:generateContent?key=$apiKey")
                .post(requestBody)
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.e(TAG, "HTTP ${response.code} — modèle: $model")
                    return null
                }
                JSONObject(response.body?.string() ?: return null)
                    .getJSONArray("candidates")
                    .getJSONObject(0)
                    .getJSONObject("content")
                    .getJSONArray("parts")
                    .getJSONObject(0)
                    .getString("text")
                    .trim()
            }
        } catch (e: Exception) {
            Log.e(TAG, "callGemini exception (model=$model)", e)
            null
        }
    }

    // ─────────────────────────────────────────────────────────
    // 1. CONTEXT READER — Texte → Meme
    // ─────────────────────────────────────────────────────────
    suspend fun generateMemeTextSuggestions(situation: String): Pair<String, String> =
        withContext(Dispatchers.IO) {
            val apiKey = getApiKey()
            if (isKeyMissing(apiKey)) {
                return@withContext "QUAND TU TRAVAILLES SANS CLÉ API" to
                    "VA SUR aistudio.google.com ET CRÉE UNE CLÉ GRATUITE !"
            }

            val prompt = """
                Tu es un générateur de memes plein d'humour d'Afrique francophone
                (Cameroun, Côte d'Ivoire, Sénégal). Tu maîtrises l'argot local et les punchlines.
                Analyse cette situation :
                "$situation"
                
                Génère un meme ultra drôle. Retourne UNIQUEMENT ce JSON brut (pas de Markdown) :
                {"top": "TEXTE DU HAUT EN MAJUSCULES", "bottom": "TEXTE DU BAS EN MAJUSCULES"}
            """.trimIndent()

            try {
                val text = callGemini(MODEL_TEXT, buildTextBody(prompt), apiKey)
                    ?: return@withContext "QUAND L'IA NE RÉPOND PAS" to "DÉSASTRE COSMIQUE"
                val obj = JSONObject(text)
                obj.optString("top", "QUAND L'IA HÉSITE") to
                obj.optString("bottom", "C'EST COMPLIQUÉ COMME SITUATION")
            } catch (e: Exception) {
                Log.e(TAG, "generateMemeTextSuggestions", e)
                "QUAND LE CODE BUG À LA LIGNE 42" to "L'EXCEPTION : ${e.localizedMessage}"
            }
        }

    // ─────────────────────────────────────────────────────────
    // 2. VOICE-TO-MEME — Audio → Transcription + Meme
    //    Gemini 2.0 Flash accepte l'audio en base64 nativement.
    //    Pas besoin de Whisper ni d'OpenAI.
    // ─────────────────────────────────────────────────────────
    suspend fun transcribeAndGenerateMeme(audioFilePath: String): AudioMemeResult =
        withContext(Dispatchers.IO) {
            val apiKey = getApiKey()
            if (isKeyMissing(apiKey)) {
                return@withContext AudioMemeResult(
                    transcription = "Clé API manquante",
                    topText       = "AUDIO REÇU SANS CLÉ API",
                    bottomText    = "VA SUR aistudio.google.com !",
                    emotion       = "confused"
                )
            }

            val file = File(audioFilePath)
            if (!file.exists() || file.length() == 0L) {
                return@withContext AudioMemeResult(
                    transcription = "",
                    topText       = "FICHIER AUDIO INTROUVABLE",
                    bottomText    = "L'ENREGISTREMENT N'A PAS FONCTIONNÉ",
                    emotion       = "sad"
                )
            }

            // Encoder le fichier audio en base64
            val audioBytes  = file.readBytes()
            val base64Audio = Base64.encodeToString(audioBytes, Base64.NO_WRAP)

            // Déterminer le mimeType
            val mimeType = when {
                audioFilePath.endsWith(".3gp")  -> "audio/3gpp"
                audioFilePath.endsWith(".m4a")  -> "audio/mp4"
                audioFilePath.endsWith(".mp3")  -> "audio/mpeg"
                audioFilePath.endsWith(".wav")  -> "audio/wav"
                audioFilePath.endsWith(".webm") -> "audio/webm"
                audioFilePath.endsWith(".ogg")  -> "audio/ogg"
                else -> "audio/3gpp"
            }

            val prompt = """
                Écoute cet enregistrement audio et :
                1. Transcris exactement ce qui est dit (en français).
                2. Identifie l'émotion principale (happy, surprised, angry, sad, confused, excited).
                3. Génère un texte de meme ultra drôle basé sur ce qui a été dit,
                   avec l'argot d'Afrique francophone si possible.
                
                Retourne UNIQUEMENT ce JSON brut (pas de Markdown, pas d'explications) :
                {
                  "transcription": "texte exact de l'audio",
                  "emotion": "surprised",
                  "top": "TEXTE DU HAUT EN MAJUSCULES",
                  "bottom": "TEXTE DU BAS EN MAJUSCULES"
                }
            """.trimIndent()

            // Corps multimodal : texte + audio inline
            val jsonBody = JSONObject().apply {
                put("contents", JSONArray().apply {
                    put(JSONObject().apply {
                        put("parts", JSONArray().apply {
                            put(JSONObject().apply { put("text", prompt) })
                            put(JSONObject().apply {
                                put("inlineData", JSONObject().apply {
                                    put("mimeType", mimeType)
                                    put("data", base64Audio)
                                })
                            })
                        })
                    })
                })
                put("generationConfig", JSONObject().apply {
                    put("responseMimeType", "application/json")
                })
            }

            try {
                val text = callGemini(MODEL_TEXT, jsonBody, apiKey)
                    ?: return@withContext AudioMemeResult(
                        transcription = "",
                        topText       = "GEMINI N'A PAS RÉPONDU",
                        bottomText    = "VÉRIFIE TA CONNEXION ET TA CLÉ API",
                        emotion       = "confused"
                    )

                val obj = JSONObject(text)
                AudioMemeResult(
                    transcription = obj.optString("transcription", ""),
                    topText       = obj.optString("top", "AUDIO REÇU ET ANALYSÉ"),
                    bottomText    = obj.optString("bottom", "MAIS L'IA N'A RIEN COMPRIS"),
                    emotion       = obj.optString("emotion", "confused")
                )
            } catch (e: Exception) {
                Log.e(TAG, "transcribeAndGenerateMeme", e)
                AudioMemeResult(
                    transcription = "",
                    topText       = "QUAND TU ENREGISTRES TA VOIX",
                    bottomText    = "ET QUE L'IA SURSAUTE EN T'ENTENDANT",
                    emotion       = "surprised"
                )
            }
        }

    /**
     * Fallback : si le fichier audio n'est pas disponible,
     * génère un meme à partir de la transcription textuelle seulement.
     */
    suspend fun generateVoiceToMemeText(audioTranscript: String): Pair<String, String> =
        withContext(Dispatchers.IO) {
            val apiKey = getApiKey()
            if (isKeyMissing(apiKey)) {
                return@withContext "AUDIO SANS CLÉ API" to "LE SILENCE EST ASSOURDISSANT"
            }

            val prompt = """
                Message vocal transcrit : "$audioTranscript"
                
                Génère un meme humoristique d'Afrique francophone.
                Retourne UNIQUEMENT ce JSON brut :
                {"top": "TEXTE DU HAUT EN MAJUSCULES", "bottom": "TEXTE DU BAS EN MAJUSCULES"}
            """.trimIndent()

            try {
                val text = callGemini(MODEL_TEXT, buildTextBody(prompt), apiKey)
                    ?: return@withContext "QUAND LA NOTE VOCALE BUG" to "PROBLÈME RÉSEAU"
                val obj = JSONObject(text)
                obj.optString("top") to obj.optString("bottom")
            } catch (e: Exception) {
                Log.e(TAG, "generateVoiceToMemeText", e)
                "QUAND TU T'ENREGISTRES" to "ET QUE TU SURSAUTES EN T'ENTENDANT"
            }
        }

    // ─────────────────────────────────────────────────────────
    // 3. STATUS REMIXER — Image → Meme (Vision multimodale)
    // ─────────────────────────────────────────────────────────
    suspend fun analyzeImageForMeme(bitmap: Bitmap): Pair<String, String> =
        withContext(Dispatchers.IO) {
            val apiKey = getApiKey()
            if (isKeyMissing(apiKey)) {
                return@withContext "PHOTO SANS CLÉ API" to "CONFIGURATION VIDE"
            }

            val out = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, 70, out)
            val base64Image = Base64.encodeToString(out.toByteArray(), Base64.NO_WRAP)

            val prompt = """
                Analyse cette image avec humour.
                Génère un texte de meme drôle en français argotique d'Afrique francophone.
                Retourne UNIQUEMENT ce JSON brut :
                {"top": "TEXTE DU HAUT EN MAJUSCULES", "bottom": "TEXTE DU BAS EN MAJUSCULES"}
            """.trimIndent()

            val jsonBody = JSONObject().apply {
                put("contents", JSONArray().apply {
                    put(JSONObject().apply {
                        put("parts", JSONArray().apply {
                            put(JSONObject().apply { put("text", prompt) })
                            put(JSONObject().apply {
                                put("inlineData", JSONObject().apply {
                                    put("mimeType", "image/jpeg")
                                    put("data", base64Image)
                                })
                            })
                        })
                    })
                })
                put("generationConfig", JSONObject().apply {
                    put("responseMimeType", "application/json")
                })
            }

            try {
                val text = callGemini(MODEL_TEXT, jsonBody, apiKey)
                    ?: return@withContext "QUAND LE MULTIMODAL CAPOTE" to "PAS DE PIXELS"
                val obj = JSONObject(text)
                obj.optString("top") to obj.optString("bottom")
            } catch (e: Exception) {
                Log.e(TAG, "analyzeImageForMeme", e)
                "L'IA QUI VOIT MA PHOTO" to "ET QUI SECOUE LA TÊTE"
            }
        }

    // ─────────────────────────────────────────────────────────
    // 4. GÉNÉRATION IMAGE — Prompt → Bitmap (Bonus)
    // ─────────────────────────────────────────────────────────
    suspend fun generateImageFromPrompt(prompt: String): Bitmap? =
        withContext(Dispatchers.IO) {
            val apiKey = getApiKey()
            if (isKeyMissing(apiKey)) return@withContext null

            val formatted = "Meme template image, funny, no text overlay, concept: $prompt. " +
                "3D cartoon style or funny vector illustration."

            val jsonBody = JSONObject().apply {
                put("contents", JSONArray().apply {
                    put(JSONObject().apply {
                        put("parts", JSONArray().apply {
                            put(JSONObject().apply { put("text", formatted) })
                        })
                    })
                })
                put("generationConfig", JSONObject().apply {
                    put("responseModalities", JSONArray().apply {
                        put("TEXT"); put("IMAGE")
                    })
                })
            }

            try {
                val requestBody = jsonBody.toString().toRequestBody("application/json".toMediaType())
                val request = Request.Builder()
                    .url("$BASE_URL/v1beta/models/$MODEL_IMAGE_GEN:generateContent?key=$apiKey")
                    .post(requestBody)
                    .build()

                client.newCall(request).execute().use { response ->
                    if (!response.isSuccessful) {
                        Log.e(TAG, "Imagen HTTP ${response.code}")
                        return@withContext null
                    }
                    val root  = JSONObject(response.body?.string() ?: return@withContext null)
                    val parts = root.getJSONArray("candidates")
                        .getJSONObject(0).getJSONObject("content").getJSONArray("parts")

                    for (i in 0 until parts.length()) {
                        val part = parts.getJSONObject(i)
                        if (part.has("inlineData")) {
                            val bytes = Base64.decode(
                                part.getJSONObject("inlineData").getString("data"),
                                Base64.DEFAULT
                            )
                            return@withContext BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                        }
                    }
                    null
                }
            } catch (e: Exception) {
                Log.e(TAG, "generateImageFromPrompt", e)
                null
            }
        }

    // ─────────────────────────────────────────────────────────
    // 5. STICKER — Contexte → Emoji + Texte court
    // ─────────────────────────────────────────────────────────
    suspend fun generateStickerSuggestion(contextText: String): Pair<String, String> =
        withContext(Dispatchers.IO) {
            val apiKey = getApiKey()
            if (isKeyMissing(apiKey)) return@withContext "🔥" to "C'EST TRÈS CHAUD !"

            val prompt = """
                Situation : "$contextText"
                Suggère un emoji expressif et une phrase max 4 mots en argot africain francophone.
                JSON brut : {"emoji": "🔥", "text": "C'EST GÂTÉ !"}
            """.trimIndent()

            try {
                val text = callGemini(MODEL_TEXT, buildTextBody(prompt), apiKey)
                    ?: return@withContext "😂" to "SAUVE QUI PEUT !"
                val obj = JSONObject(text)
                obj.optString("emoji", "😂") to obj.optString("text", "ON EST ENSEMBLE !")
            } catch (e: Exception) {
                Log.e(TAG, "generateStickerSuggestion", e)
                "💀" to "FIN DU GAMING !"
            }
        }

    // ─────────────────────────────────────────────────────────
    // 6. GIF SEARCH — Contexte → Requête de recherche
    // ─────────────────────────────────────────────────────────
    suspend fun generateGifSearchQuery(contextText: String): String =
        withContext(Dispatchers.IO) {
            val apiKey = getApiKey()
            if (isKeyMissing(apiKey)) return@withContext "funny cat screaming"

            val prompt = """
                Contexte : "$contextText"
                Requête GIF en anglais, courte et drôle.
                JSON brut : {"query": "excited minion jumping"}
            """.trimIndent()

            try {
                val text = callGemini(MODEL_TEXT, buildTextBody(prompt), apiKey)
                    ?: return@withContext "funny face"
                JSONObject(text).optString("query", "funny face")
            } catch (e: Exception) {
                Log.e(TAG, "generateGifSearchQuery", e)
                "funny facepalm reaction"
            }
        }

    // ─────────────────────────────────────────────────────────
    // 7. VIDEO STORYBOARD — Contexte → Titre + Punchline
    // ─────────────────────────────────────────────────────────
    suspend fun generateVideoStoryboard(contextText: String): Pair<String, String> =
        withContext(Dispatchers.IO) {
            val apiKey = getApiKey()
            if (isKeyMissing(apiKey)) return@withContext "PROJET EN FEU" to "C'EST GÂTÉ !"

            val prompt = """
                Contexte : "$contextText"
                Titre 2-3 mots + punchline rythmée pour une vidéo TikTok/Reels.
                JSON brut : {"title": "ALERTE CODEUR", "punchline": "QUAND LE CLIENT VEUT DU 3D DEMAIN"}
            """.trimIndent()

            try {
                val text = callGemini(MODEL_TEXT, buildTextBody(prompt), apiKey)
                    ?: return@withContext "STORY" to "C'EST CHAUD !"
                val obj = JSONObject(text)
                obj.optString("title", "ALERTE") to obj.optString("punchline", "DÉSASTRE")
            } catch (e: Exception) {
                Log.e(TAG, "generateVideoStoryboard", e)
                "BUG REPORT" to "L'EXCEPTION A ENCORE FRAPPÉ !"
            }
        }
}

// ─────────────────────────────────────────────────────────────
// Modèle de données pour le résultat audio
// ─────────────────────────────────────────────────────────────
data class AudioMemeResult(
    val transcription : String,
    val topText       : String,
    val bottomText    : String,
    val emotion       : String
)

File: meme-multimodal-V2/src/.OLD FILES/api/GeminiClient.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

const BASE_URL_GEMINI = 'https://generativelanguage.googleapis.com';
const BASE_URL_EDEN = 'https://api.edenai.run/v2';

/**
 * Clean up text response and extract valid JSON block.
 * This is 100% robust against markdown backticks (```json ... ```) or extra prose.
 */
function cleanAndExtractJson(input: string): string {
  let str = input.trim();
  
  if (str.startsWith('```')) {
    const lastBackticks = str.lastIndexOf('```');
    if (lastBackticks > 3) {
      const firstLineEnd = str.indexOf('\n');
      if (firstLineEnd !== -1 && firstLineEnd < lastBackticks) {
        str = str.substring(firstLineEnd + 1, lastBackticks);
      } else {
        str = str.replace(/```json/g, '').replace(/```/g, '');
      }
    } else {
      str = str.replace(/```json/g, '').replace(/```/g, '');
    }
  }
  
  str = str.trim();
  
  const firstBrace = str.indexOf('{');
  const lastBrace = str.lastIndexOf('}');
  if (firstBrace !== -1 && lastBrace !== -1 && lastBrace > firstBrace) {
    return str.substring(firstBrace, lastBrace + 1);
  }
  
  const firstBracket = str.indexOf('[');
  const lastBracket = str.lastIndexOf(']');
  if (firstBracket !== -1 && lastBracket !== -1 && lastBracket > firstBracket) {
    return str.substring(firstBracket, lastBracket + 1);
  }
  
  return str;
}

export const GeminiClient = {
  /**
   * Retrieves keys from local AsyncStorage storage or fallback environment variables.
   */
  getApiKeys: async (): Promise<{ gemini: string; eden: string }> => {
    try {
      const gemini = await AsyncStorage.getItem('GEMINI_API_KEY');
      const eden = await AsyncStorage.getItem('EDEN_API_KEY');
      return {
        gemini: gemini || '',
        eden: eden || ''
      };
    } catch {
      return { gemini: '', eden: '' };
    }
  },

  /**
   * 1. Generates a pair of texts (top + bottom) for a situation context
   */
  generateMemeTextSuggestions: async (situation: string): Promise<{ top: string; bottom: string }> => {
    const keys = await GeminiClient.getApiKeys();
    
    const prompt = `
      Tu es un générateur de memes plein d'humour et d'ironie d'Afrique de l'Ouest (Cameroun, Côte d'Ivoire, Sénégal...), fan de blagues courtes et de punchlines locales.
      Analyse la situation ou discussion suivante :
      "${situation}"
      
      Suggère un texte de meme ultra drôle adapté à cette situation. 
      Retourne STRICTEMENT un objet JSON avec deux propriétés 'top' et 'bottom' (toutes deux en lettres MAJUSCULES).
      IMPORTANT: Ne mets aucun bloc de code Markdown, juste le JSON brut valide. Pas d'explications superflues.
      Format attendu :
      {"top": "TEXTE DU HAUT EN MAJUSCULES", "bottom": "TEXTE DU BAS EN MAJUSCULES"}
    `.trim();

    // Use direct Gemini if available
    if (keys.gemini && keys.gemini !== 'MY_GEMINI_API_KEY') {
      try {
        const response = await fetch(`${BASE_URL_GEMINI}/v1beta/models/gemini-3.5-flash:generateContent?key=${keys.gemini}`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: { responseMimeType: 'application/json' }
          })
        });

        if (response.ok) {
          const resJson = await response.json();
          const rawText = resJson.candidates[0].content.parts[0].text;
          const cleanJsonStr = cleanAndExtractJson(rawText);
          const parsed = JSON.parse(cleanJsonStr);
          return {
            top: (parsed.top || "QUAND L'IA ESSAYE DE FAIRE TOUT SEULE").toUpperCase(),
            bottom: (parsed.bottom || "C'EST DRÔLE MAIS PEUT MIEUX FAIRE").toUpperCase()
          };
        }
      } catch (e) {
        console.warn('Direct Gemini call failed in generateMemeTextSuggestions, using fallback...', e);
      }
    }

    // Fallback to Eden AI
    if (keys.eden && keys.eden !== 'MY_EDEN_API_KEY') {
      try {
        const response = await fetch(`${BASE_URL_EDEN}/text/generation`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${keys.eden}`
          },
          body: JSON.stringify({
            providers: 'google',
            text: prompt,
            temperature: 0.7,
            max_tokens: 300
          })
        });

        if (response.ok) {
          const resJson = await response.json();
          const rawText = resJson.google.generated_text;
          const cleanJsonStr = cleanAndExtractJson(rawText);
          const parsed = JSON.parse(cleanJsonStr);
          return {
            top: (parsed.top || "QUAND EDEN AI VIENT À LA RESCOUSSE").toUpperCase(),
            bottom: (parsed.bottom || "TOUT REPART SANS ENCOMBRES !").toUpperCase()
          };
        }
      } catch (e) {
        console.error('Eden AI generateMemeTextSuggestions failed:', e);
      }
    }

    // Default simulation fallback
    return {
      top: "QUAND LA CLÉ API MANQUE À L'APPEL",
      bottom: "SOUFFLE UN COUP ET METS À JOUR TES CLÉS DANS LES RÉGLAGES !"
    };
  },

  /**
   * 2. Voice-to-Meme suggestions based on audio transcript
   */
  generateVoiceToMemeText: async (audioTranscript: string): Promise<{ top: string; bottom: string }> => {
    const keys = await GeminiClient.getApiKeys();
    const prompt = `
      Analyse la transcription de note vocale suivante : "${audioTranscript}"
      Génère un meme textuel hilarant. Détecte le sentiment ou la blague.
      Retourne STRICTEMENT un objet JSON avec 'top' et 'bottom' en lettres MAJUSCULES.
      Format attendu :
      {"top": "TEXTE DU HAUT", "bottom": "TEXTE DU BAS"}
    `.trim();

    if (keys.gemini && keys.gemini !== 'MY_GEMINI_API_KEY') {
      try {
        const response = await fetch(`${BASE_URL_GEMINI}/v1beta/models/gemini-3.5-flash:generateContent?key=${keys.gemini}`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: { responseMimeType: 'application/json' }
          })
        });

        if (response.ok) {
          const resJson = await response.json();
          const rawText = resJson.candidates[0].content.parts[0].text;
          const cleanJsonStr = cleanAndExtractJson(rawText);
          const parsed = JSON.parse(cleanJsonStr);
          return {
            top: (parsed.top || "QUAND TU ENVOIES UN VOCAL SANS PRESSION").toUpperCase(),
            bottom: (parsed.bottom || "ET QUE L'IA DÉBALLE TOUT TON SECRET !").toUpperCase()
          };
        }
      } catch {}
    }

    // Default simulation fallback
    return {
      top: "QUAND LA TRANSCRIPTION SE TERMINE EN BEAUTÉ",
      bottom: "ET QUE LE GROUPE DE DISCUSSION RESTE SANS VOIX !"
    };
  },

  /**
   * 3. Analyze image in base64 (multimodal) to generate meme overlays
   */
  analyzeImageForMeme: async (imageBase64: string, mimeType: string = 'image/jpeg'): Promise<{ top: string; bottom: string }> => {
    const keys = await GeminiClient.getApiKeys();
    const prompt = `
      Analyse cette image. Décris brièvement la situation avec humour et suggère un texte pour en faire un meme.
      Retourne STRICTEMENT un objet JSON brut valide avec deux clés 'top' et 'bottom', rédigés en français / argot jeune percutant.
      Format attendu :
      {"top": "TEXTE DU HAUT EN MAJUSCULES", "bottom": "TEXTE DU BAS EN MAJUSCULES"}
    `.trim();

    if (keys.gemini && keys.gemini !== 'MY_GEMINI_API_KEY') {
      try {
        const pureBase64 = imageBase64.includes('base64,') ? imageBase64.split('base64,')[1] : imageBase64;
        const response = await fetch(`${BASE_URL_GEMINI}/v1beta/models/gemini-3.5-flash:generateContent?key=${keys.gemini}`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            contents: [{
              parts: [
                { text: prompt },
                { inlineData: { mimeType, data: pureBase64 } }
              ]
            }],
            generationConfig: { responseMimeType: 'application/json' }
          })
        });

        if (response.ok) {
          const resJson = await response.json();
          const rawText = resJson.candidates[0].content.parts[0].text;
          const cleanJsonStr = cleanAndExtractJson(rawText);
          const parsed = JSON.parse(cleanJsonStr);
          return {
            top: (parsed.top || "L'IA DÉCOUVRE TON ÉCRAN").toUpperCase(),
            bottom: (parsed.bottom || "ET COMPREND DIRECTEMENT TON CODE DE L'ENFER").toUpperCase()
          };
        }
      } catch (e) {
        console.warn('Multimodal Direct Gemini failed:', e);
      }
    }

    return {
      top: "QUAND L'IMAGE EST TROP PUISSANTE",
      bottom: "SANS CONFIGURATION DE CLÉ MULTIMODALE"
    };
  },

  /**
   * 4. Generates a custom background using Imagen model (gemini-2.5-flash-image)
   */
  generateImageFromPrompt: async (prompt: string): Promise<string | null> => {
    const keys = await GeminiClient.getApiKeys();
    if (!keys.gemini || keys.gemini === 'MY_GEMINI_API_KEY') return null;

    const formattedPrompt = `Meme template raw image, highly expressive, concept: ${prompt}. Funny background without any text. 3D cartoon style or funny vector design.`;

    try {
      const response = await fetch(`${BASE_URL_GEMINI}/v1beta/models/gemini-2.5-flash-image:generateContent?key=${keys.gemini}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          contents: [{ parts: [{ text: formattedPrompt }] }],
          generationConfig: {
            responseModalities: ["TEXT", "IMAGE"],
            imageConfig: {
              aspectRatio: "1:1",
              imageSize: "512"
            }
          }
        })
      });

      if (response.ok) {
        const resJson = await response.json();
        const parts = resJson.candidates[0].content.parts;
        for (let i = 0; i < parts.length; i++) {
          if (parts[i].inlineData) {
            const base64Data = parts[i].inlineData.data;
            const mimeType = parts[i].inlineData.mimeType || 'image/png';
            return `data:${mimeType};base64,${base64Data}`;
          }
        }
      }
    } catch (e) {
      console.error('Error generating image via Imagen:', e);
    }
    return null;
  },

  /**
   * 5. Suggère un emoji et un texte court de sticker adaptés au contexte donné.
   */
  generateStickerSuggestion: async (contextText: string): Promise<{ emoji: string; text: string }> => {
    const keys = await GeminiClient.getApiKeys();
    const prompt = `
      On a cette situation ou contexte : "${contextText}"
      Suggère un emoji unique très expressif (comme 😂, 💀, 🤡, 🚀, 🤦‍♂️, 🤯) et une phrase ultra-courte (max 3-4 mots, en français argotique jeune et marrant d'Afrique francophone de préférence) pour en faire un sticker humoristique.
      Retourne STRICTEMENT un objet JSON brut avec les clés 'emoji' et 'text'.
      Format attendu :
      {"emoji": "🔥", "text": "C'EST GÂTÉ !"}
    `.trim();

    if (keys.gemini && keys.gemini !== 'MY_GEMINI_API_KEY') {
      try {
        const response = await fetch(`${BASE_URL_GEMINI}/v1beta/models/gemini-3.5-flash:generateContent?key=${keys.gemini}`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: { responseMimeType: 'application/json' }
          })
        });

        if (response.ok) {
          const resJson = await response.json();
          const rawText = resJson.candidates[0].content.parts[0].text;
          const cleanJsonStr = cleanAndExtractJson(rawText);
          const parsed = JSON.parse(cleanJsonStr);
          return {
            emoji: parsed.emoji || '😂',
            text: (parsed.text || 'ON EST ENSEMBLE !').toUpperCase()
          };
        }
      } catch {}
    }

    return { emoji: '🔥', text: "C'EST TRÈS CHAUD !" };
  },

  /**
   * 6. Generates a GIF search query
   */
  generateGifSearchQuery: async (contextText: string): Promise<string> => {
    const keys = await GeminiClient.getApiKeys();
    const prompt = `
      On a ce contexte : "${contextText}"
      Suggère une requête de recherche de GIF en anglais très précise et drôle décrivant cette situation exacte (par exemple "confused programmer crying", "excited baby dancing", "awkward stare facepalm").
      Retourne STRICTEMENT un objet JSON brut avec une clé 'query' sans balises Markdown.
      Format attendu :
      {"query": "excited minion jumping"}
    `.trim();

    if (keys.gemini && keys.gemini !== 'MY_GEMINI_API_KEY') {
      try {
        const response = await fetch(`${BASE_URL_GEMINI}/v1beta/models/gemini-3.5-flash:generateContent?key=${keys.gemini}`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: { responseMimeType: 'application/json' }
          })
        });

        if (response.ok) {
          const resJson = await response.json();
          const rawText = resJson.candidates[0].content.parts[0].text;
          const cleanJsonStr = cleanAndExtractJson(rawText);
          const parsed = JSON.parse(cleanJsonStr);
          return parsed.query || 'funny cat screaming';
        }
      } catch {}
    }

    return 'funny cat screaming';
  },

  /**
   * 7. Generates animated Short Video storyboard title and punchline
   */
  generateVideoStoryboard: async (contextText: string): Promise<{ title: string; punchline: string }> => {
    const keys = await GeminiClient.getApiKeys();
    const prompt = `
      On a cette situation : "${contextText}"
      Imagine un storyboard court pour une vidéo TikTok/Reel humoristique de 5 secondes.
      Suggère un titre d'introduction dramatique ('title') et une punchline finale fracassante ('punchline').
      Retourne STRICTEMENT un objet JSON brut avec ces clés sans fioritures.
      Format attendu :
      {"title": "ALERTE MAXIMALE", "punchline": "DÉSASTRE GÉNÉRALISÉ"}
    `.trim();

    if (keys.gemini && keys.gemini !== 'MY_GEMINI_API_KEY') {
      try {
        const response = await fetch(`${BASE_URL_GEMINI}/v1beta/models/gemini-3.5-flash:generateContent?key=${keys.gemini}`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            contents: [{ parts: [{ text: prompt }] }],
            generationConfig: { responseMimeType: 'application/json' }
          })
        });

        if (response.ok) {
          const resJson = await response.json();
          const rawText = resJson.candidates[0].content.parts[0].text;
          const cleanJsonStr = cleanAndExtractJson(rawText);
          const parsed = JSON.parse(cleanJsonStr);
          return {
            title: (parsed.title || 'ALERTE DRAME').toUpperCase(),
            punchline: (parsed.punchline || 'VRAIMENT UNE HISTOIRE DE FOU').toUpperCase()
          };
        }
      } catch {}
    }

    return {
      title: 'ALERTE MAXIMALE ⚠️',
      punchline: 'DÉSASTRE GÉNÉRALISÉ ET IRRÉVERSIBLE !'
    };
  }
};
